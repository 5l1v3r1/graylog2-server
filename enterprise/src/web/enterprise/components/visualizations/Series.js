import Immutable from 'immutable';

const _defaultSeriesGenerator = (type, name, labels, values) => ({ type, name, x: labels, y: values });
export const generateSeries = (config, results, chartType, generator = _defaultSeriesGenerator) => {
  const { series } = config;
  const rowPivots = config.rowPivots.map(({ field }) => field);
  const columnPivots = config.columnPivots.map(({ field }) => field);
  const x = results.map(v => rowPivots.map(p => v[p]).join('-'));

  const columnPivotCharts = columnPivots.map((columnPivot) => {
    const values = {};
    const columnPivotX = results.filter(r => r[columnPivot])
      .map(row => row[columnPivot].map(pivot => pivot[columnPivot]))
      .reduce((prev, cur) => prev.merge(cur), Immutable.OrderedSet()).toJS();
    results.forEach((row) => {
      const valueKey = rowPivots.map(p => row[p]).join('-');
      columnPivotX.forEach((columnPivotValue) => {
        series.forEach((seriesName) => {
          const pivotValueKey = `${columnPivotValue}-${seriesName}`;

          if (!values[pivotValueKey]) {
            values[pivotValueKey] = [];
          }

          const rowValue = (row[columnPivot].find(r => r[columnPivot] === columnPivotValue) || {})[seriesName];
          values[`${columnPivotValue}-${seriesName}`].push({ key: valueKey, value: rowValue });
        });
      });
    });

    return Object.keys(values).map(pivotValue => [
      chartType,
      pivotValue,
      values[pivotValue].map(({ key }) => key),
      values[pivotValue].map(({ value }) => value),
    ]);
  }).flatten();

  const seriesCharts = series.map(s => [chartType, s, x, results.map(v => v[s])]);

  const allCharts = [].concat(seriesCharts, columnPivotCharts);
  return allCharts.map((args, idx) => generator(...args, idx, allCharts.length));
};
