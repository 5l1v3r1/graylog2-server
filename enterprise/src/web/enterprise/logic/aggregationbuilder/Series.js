import { get } from 'lodash';
import * as Immutable from 'immutable';

import SeriesConfig from './SeriesConfig';

export default class Series {
  constructor(func, config = SeriesConfig.empty()) {
    this._value = { function: func, config };
  }

  get function() {
    return this._value.function;
  }

  get config() {
    return this._value.config;
  }

  get effectiveName() {
    const overridenName = get(this, 'config.name');
    return overridenName || this.function;
  }

  toString() {
    return `Series: ${this.effectiveName}, config={${this.config.toJSON()}}`;
  }

  toJSON() {
    return {
      config: this._value.config,
      function: this._value.function,
    };
  }

  static fromJSON(value) {
    return new Series(value.function, SeriesConfig.fromJSON(value.config));
  }

  static forFunction(func) {
    // eslint-disable-next-line no-use-before-define
    return new Builder().function(func).config(SeriesConfig.empty()).build();
  }

  toBuilder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map(this._value));
  }
}

class Builder {
  constructor(value = Immutable.Map()) {
    this.value = value;
  }

  function(newFunction) {
    return new Builder(this.value.set('function', newFunction));
  }

  config(newConfig) {
    return new Builder(this.value.set('config', newConfig));
  }

  build() {
    const { config } = this.value.toObject();
    const func = this.value.get('function');
    return new Series(func, config);
  }
}
