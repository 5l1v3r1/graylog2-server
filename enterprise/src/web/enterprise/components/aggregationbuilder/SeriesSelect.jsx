import React from 'react';
import PropTypes from 'prop-types';

import Select from 'react-select';
import ConfigurableElement from './ConfigurableElement';
import SeriesConfiguration from './SeriesConfiguration';
import SeriesFunctionsSuggester from './SeriesFunctionsSuggester';
import Series from '../../logic/aggregationbuilder/Series';

const parseSeries = series => (series ? series.map(s => s.value) : []);

const newSeriesConfigChange = (values, series, newSeries, onChange) => {
  const newValues = values.map(s => (s === series ? newSeries : s));
  return onChange(newValues);
};

const _wrapOption = series => ({ label: series.effectiveName, value: series });

class SeriesSelect extends React.Component {
  constructor(props, context) {
    super(props, context);
    const { suggester } = props;
    this.state = {
      options: suggester.defaults,
    };
  }
  _onChange = (newSeries) => {
    const last = newSeries[newSeries.length - 1];
    if (!last) {
      return false;
    }

    const { incomplete, backToFunctions, value } = last;

    if (incomplete) {
      const options = [].concat(
        [{ label: 'Back to function list', backToFunctions: true }],
        this.props.suggester.for(value),
      );
      this.setState({ options });
      return false;
    }

    this._resetToFunctions();

    if (backToFunctions) {
      return false;
    }

    this.props.onChange(parseSeries(newSeries));
    return true;
  };

  _resetToFunctions = () => {
    this.setState({ options: this.props.suggester.defaults });
  };

  _onClose = () => {
    this._resetToFunctions();
  };

  render() {
    const { onChange, series } = this.props;
    const valueComponent = ({ children, ...rest }) => {
      const element = rest.value.value;
      return (
        <ConfigurableElement {...rest}
                             configuration={({ onClose }) => <SeriesConfiguration series={element} onClose={onClose} />}
                             onChange={newElement => newSeriesConfigChange(series, element, newElement, onChange)}
                             title="Series Configuration">
          {children}
        </ConfigurableElement>
      );
    };
    return (<Select placeholder="None: click to add series"
                    onChange={this._onChange}
                    options={this.state.options}
                    value={series.map(_wrapOption)}
                    valueComponent={valueComponent}
                    onClose={this._onClose}
                    ignoreCase
                    closeOnSelect={false}
                    openOnClick
                    openOnFocus
                    onBlurResetsInput
                    onCloseResetsInput
                    scrollMenuIntoView
                    tabSelectsValue
                    escapeClearsValue
                    multi />);
  }
}

SeriesSelect.propTypes = {
  onChange: PropTypes.func.isRequired,
  series: PropTypes.arrayOf(PropTypes.instanceOf(Series)).isRequired,
  suggester: PropTypes.any,
};

SeriesSelect.defaultProps = {
  suggester: new SeriesFunctionsSuggester(),
};

export default SeriesSelect;
