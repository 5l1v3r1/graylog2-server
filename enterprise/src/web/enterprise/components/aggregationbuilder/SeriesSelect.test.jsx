import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import SeriesSelect from './SeriesSelect';
import Series from '../../logic/aggregationbuilder/Series';

jest.mock('enterprise/stores/AggregationFunctionsStore', () => ({ getInitialState: jest.fn(), listen: jest.fn() }));

describe('SeriesSelect', () => {
  it('renders with minimal props', () => {
    const wrapper = renderer.create(<SeriesSelect series={[]} onChange={() => {}} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders given series', () => {
    const series = [Series.forFunction('count()'), Series.forFunction('avg(took_ms)')];
    const wrapper = mount(<SeriesSelect series={series} onChange={() => {}} />);

    expect(wrapper.find('Value')).toHaveLength(2);
    const values = wrapper.find('Value');
    expect(values.at(0)).toIncludeText('count()');
    expect(values.at(1)).toIncludeText('avg(took_ms)');
  });

  it('opens menu when focussed, returning no results without suggester', () => {
    const wrapper = mount(<SeriesSelect series={[]} onChange={() => {}} />);
    const input = wrapper.find('input');
    expect(wrapper).not.toIncludeText('No results found');

    input.simulate('focus');

    expect(wrapper).toIncludeText('No results found');
  });

  it('opens menu when focussed, returning no results for empty suggester defaults', () => {
    const suggester = { defaults: [] };
    const wrapper = mount(<SeriesSelect series={[]} suggester={suggester} onChange={() => {}} />);
    const input = wrapper.find('input');
    expect(wrapper).not.toIncludeText('No results found');

    input.simulate('focus');

    expect(wrapper).toIncludeText('No results found');
  });

  it('opens menu when focussed, returning suggester defaults', () => {
    const suggester = { defaults: [{ label: 'Something', value: 'Something' }, { label: 'Anything', value: 'Anything' }] };
    const wrapper = mount(<SeriesSelect series={[]} suggester={suggester} onChange={() => {}} />);
    const input = wrapper.find('input');
    expect(wrapper).not.toIncludeText('Something');
    expect(wrapper).not.toIncludeText('Anything');

    input.simulate('focus');

    expect(wrapper).not.toIncludeText('No results found');
    expect(wrapper).toIncludeText('Something');
    expect(wrapper).toIncludeText('Anything');
  });

  it('shows next suggestions when selecting incomplete suggestion', () => {
    const suggester = {
      defaults: [{ label: 'func1', value: 'func1', incomplete: true }, { label: 'Anything', value: 'Anything' }],
      for: jest.fn(() => [
        { label: 'func1(foo)', value: 'func1(foo)' },
        { label: 'func1(bar)', value: 'func1(bar)' },
      ]),
    };
    const onChange = jest.fn();
    const wrapper = mount(<SeriesSelect series={[]} suggester={suggester} onChange={onChange} />);
    const input = wrapper.find('input');
    input.simulate('focus');

    const select = wrapper.find('Select');
    select.prop('onChange')([{ label: 'func1', value: 'func1', incomplete: true }]);

    expect(suggester.for).toHaveBeenCalledTimes(1);
    expect(suggester.for).toHaveBeenCalledWith('func1');
    expect(onChange).not.toHaveBeenCalled();
  });

  it('unwraps value and returns it when final option was selected', () => {
    const suggester = {};
    const onChange = jest.fn();
    const wrapper = mount(<SeriesSelect series={[]} suggester={suggester} onChange={onChange} />);

    expect(onChange).not.toHaveBeenCalled();

    const select = wrapper.find('Select');
    select.prop('onChange')([{ label: 'func1(foo)', value: 'func1(foo)' }]);

    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith(['func1(foo)']);
  });
});
