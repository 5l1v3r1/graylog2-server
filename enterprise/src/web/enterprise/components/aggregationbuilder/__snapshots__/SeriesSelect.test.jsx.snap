// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesSelect renders with minimal props 1`] = `
<div
  className="Select is-clearable is-searchable Select--multi"
>
  <div
    className="Select-control"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="Select-multi-value-wrapper"
      id="react-select-2--value"
    >
      <div
        className="Select-placeholder"
      >
        None: click to add series
      </div>
      <div
        className="Select-input"
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          aria-activedescendant="react-select-2--value"
          aria-expanded="false"
          aria-haspopup="false"
          aria-owns=""
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          role="combobox"
          style={
            Object {
              "boxSizing": "content-box",
              "width": "5px",
            }
          }
          value=""
        />
        <div
          style={
            Object {
              "height": 0,
              "left": 0,
              "overflow": "scroll",
              "position": "absolute",
              "top": 0,
              "visibility": "hidden",
              "whiteSpace": "pre",
            }
          }
        >
          
        </div>
      </div>
    </span>
    <span
      className="Select-arrow-zone"
      onMouseDown={[Function]}
    >
      <span
        className="Select-arrow"
        onMouseDown={[Function]}
      />
    </span>
  </div>
</div>
`;
